components/ContainedInputViews/src/MDCContainedInputView.h

```diff
@@ -0,0 +1,319 @@
+  /**
+   The error state.
+   */
+  MDCContainedInputViewStateErrored = 1 << 3,
+  /**
+   The disabled state.
+   */
+  MDCContainedInputViewStateDisabled = 1 << 4,
+};
+
+/**
+ Dictates the relative importance of the underline labels, and the order in which they are laid out.
+ */
+typedef NS_ENUM(NSUInteger, MDCContainedInputViewUnderlineLabelDrawPriority) {
+  /**
+   When the priority is @c .leading, the @c leadingUnderlineLabel will be laid out first within the
+   horizontal space available for @b both underline labels. Any remaining space will then be given
+   for the @c trailingUnderlineLabel.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityLeading,
+  /**
+   When the priority is @c .trailing, the @c trailingUnderlineLabel will be laid out first within
+   the horizontal space available for @b both underline labels. Any remaining space will then be
+   given for the @c leadingUnderlineLabel.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityTrailing,
+  /**
+   When the priority is @c .custom, the @c customUnderlineLabelDrawPriority property will be used to
+   divide the space available for the two underline labels.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityCustom,
+};
+
+/**
+ This enum represents different states the floating label can be in.
+ */
+typedef NS_ENUM(NSUInteger, MDCContainedInputViewFloatingLabelState) {
+  /**
+   The state where the floating label is not visible.
+   */
+  MDCContainedInputViewFloatingLabelStateNone,
+  /**
+   The state where the floating label is floating.
+   */
+  MDCContainedInputViewFloatingLabelStateFloating,
+  /**
+   The state where the floating label is occupying the normal text area.
+   */
+  MDCContainedInputViewFloatingLabelStateNormal,
+};
+
+@protocol MDCContainedInputViewStyle;
+@protocol MDCContainedInputViewColorScheming;
+
+@protocol MDCContainedInputView <NSObject>
+/**
+ Dictates the @c MDCContainerStyle of the text field. Defaults to an instance of
+ MDCContainerStyleBase.
+ */
+@property(nonatomic, strong, nonnull) id<MDCContainedInputViewStyle> containerStyle;
+
+/**
+ Describes the current @c MDCContainedInputViewState of the view.
+ */
+@property(nonatomic, assign, readonly) MDCContainedInputViewState containedInputViewState;
+
+/**
+ Describes the current @c MDCContainedInputViewFloatingLabelState of the contained input view. This
+ value is affected by things like the view's state, the value for @c canFloatingLabelFloat, and the
+ text of the floating label.
+ */
+@property(nonatomic, assign, readonly) MDCContainedInputViewFloatingLabelState floatingLabelState;
+
+/**
+ The @c floatingLabel is a label that occupies the text area when there is no text and that floats
+ above the text once there is some. It is distinct from a placeholder.
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *floatingLabel;
+/**
+ The @c leadingUnderlineLabel can be used to display helper or error text.
```

Comment:
Tip: If this is its purpose, perhaps it should be part of the name? Like helperLabel? underlineLabel seems a little vague.
Tip: Add a usage examples for the labels, so it's clear to users how they should use them.

Reply:
I carried this over from MDCTextField. I think the idea is that if it can be helper text or error text it doesn't make sense to call it either of those. Simply calling it an underline label is less limiting.